docs/internal/future_optimisations.md

```diff
@@ -0,0 +1,163 @@
+The first case may result in significant code-size increases and so is worth doing only after applying a cost model.
+
+The combination of this and descriptor elision should make it possible to have efficient unit types.
+For example, consider something like this trivial example:
+
+```verona
+class Seconds
+{
+	var _value : U64;
+	+(self: Self & mut, inc : U64)
+	{
+		_value = inc;
+	}
+}
+```
+
+In the MVP compiler, this will require three words (the region metadata, the descriptor, and the 64-bit integer).
+On a CHERI system, this ends up being 40 bytes to store an 8-byte quantity.
+Now consider the following (assuming `Nanoseconds` has the same layout as `Seconds`):
+
+```verona
+class TimeInterval
+{
+	embed _seconds     : Seconds;
+	embed _nanoseconds : Nanoseconds;
+	// Methods here.
+	<(self : Self, other : TimeInterval) { ... }
+}
+```
+
+This class stores 16 bytes of state but would require 96 bytes in the MVP compiler (on a non-CHERI 64-bit system).
+Both `_seconds` and `_nanoseconds` are used by only this class and so we should be able to elide both of their descriptors and region metadata fields, shrinking this class to 32 bytes.
+If `TimeInterval` is never used in an interface or union type, then we can shrink it to 24 bytes.
+If nothing uses `TimeInterval` for region identification then we can shrink `embed` fields of type `TimeInterval` to 16 bytes.
+
+Note that it is not possible to remove the region metadata for objects that are not allocated within other objects in the general case because it is needed for memory management.
+It may be possible to elide region metadata for objects if all following hold:
+
+ - They are allocated in arena regions.
+ - They do not have a finaliser.
+ - Nothing uses objects of this type to dynamically identify a region.
+
+The first two means that there is no need to track state for memory management.
+The objects remain life until the region is deallocated and are then cleaned up by a bulk free operation.
+The last means that nothing outside of the memory manager needs to inspect the region metadata.
+
+As a further optimisation, if another object from the same region is visible in all callers of the function that access the region and the first two conditions are satisfied, then an additional region parameter could be added to the functions to propagate the region information.
+
+Tagged unions for closed-world types
+------------------------------------
+
+A union type with a small number of concrete types (for example `A | B`) will be represented in the MVP compiler as an interface that contains all of the methods in `A` and all of the methods in `B` (and possibly some others generated by the compiler to support later casts).
+Casting back to `A` or `B` requires comparing the descriptor pointer.
+This requires materialising the global addresses of the descriptors and loading from the object.
+
+We could alternatively represent these using the low bit(s) as a discriminator.
```

Comment:
I thought we were using the low bits for snmalloc purposes.

If not, then we could only apply this for small closed-world unions (up to 4).